src/Components/BehaviorsModal/Internal/Widgets/DataHistoryWidget/DataHistoryWidget.tsx

```diff
@@ -1,47 +1,178 @@
-import React, { useContext } from 'react';
-import { IDataHistoryWidget } from '../../../../../Models/Types/Generated/3DScenesConfiguration-v1.0.0';
+import React, { memo, useContext, useEffect, useMemo, useRef } from 'react';
+import {
+    ADXTimeSeries,
+    BehaviorModalMode,
+    DTwin,
+    IDataHistoryWidgetTimeSeriesTwin,
+    TimeSeriesData
+} from '../../../../../Models/Constants';
+import { useTimeSeriesData } from '../../../../../Models/Hooks/useTimeSeriesData';
+import {
+    IDataHistoryTimeSeries,
+    IDataHistoryWidget,
+    IDataHistoryWidgetConfiguration
+} from '../../../../../Models/Types/Generated/3DScenesConfiguration-v1.0.0';
+import HighChartsWrapper from '../../../../HighChartsWrapper/HighChartsWrapper';
+import { IHighChartSeriesData } from '../../../../HighChartsWrapper/HighChartsWrapper.types';
 import { BehaviorsModalContext } from '../../../BehaviorsModal';
+import { dataHistoryClassNames } from './DataHistoryWidget.styles';
 
 interface IProp {
     widget: IDataHistoryWidget;
 }
 
-export const DataHistoryWidget: React.FC<IProp> = ({ widget }) => {
-    const { mode } = useContext(BehaviorsModalContext);
+const DataHistoryWidget: React.FC<IProp> = ({ widget }) => {
     const {
         displayName,
         connectionString,
         timeSeries,
         chartOptions
     } = widget.widgetConfiguration;
 
+    const { adapter, twins, mode } = useContext(BehaviorsModalContext);
+    const twinIdPropertyMap = getTwinIdPropertyMap(timeSeries, twins);
+    const isRequestSent = useRef(false);
+
+    const {
+        query,
+        deeplink,
+        data,
+        fetchTimeSeriesData,
+        isLoading
+    } = useTimeSeriesData({
+        adapter,
+        connectionString,
+        quickTimeSpan: chartOptions.defaultQuickTimeSpan,
+        twins: twinIdPropertyMap
+    });
+
+    useEffect(() => {
+        if (
+            mode === BehaviorModalMode.viewer &&
+            query &&
+            (adapter || connectionString) &&
+            !isRequestSent.current &&
+            twinIdPropertyMap
+        ) {
+            fetchTimeSeriesData();
+            isRequestSent.current = true;
+        }
+    }, [
+        adapter,
+        query,
+        connectionString,
+        twinIdPropertyMap,
+        chartOptions,
+        mode
+    ]);
+
+    const placeholderTimeSeriesData: Array<
+        Array<TimeSeriesData>
+    > = useMemo(() => {
+        // placeholder timeseries data to be used in preview mode, need to memoize not to change the chart plot area unless time span or series length changes
+        const toInMillis = Date.now();
```

Comment:
this method is only used by this component now, if we need to import it by any other component we can move it to utils

Reply:
now I needed to use it, so moved it to utils

Reply:
I meant pull it out of the render function to a module scope function. Ideally it would not be in the general utils file as it will never be used across other components. IMO the fewer things we put at the global level the better since most htings are not actually that generic.